### PR TITLE
Fix PPO log_ratio bug

### PR DESCRIPTION
**Pull request by TobiasNorlund**
Relevant issue: #508 

- Set `position_ids` when computing logprobs, both in `make_experience` and in `loss` to ensure same absolute positional embeddings are used in the two methods.
- I _think_ the mask applied when computing `logratio` should be shifted by one, to correctly mask the last token in the batch.

**Note**: Remember to remove the debug print statements before merge.

### CODE REVIEW

**Comment by TobiasNorlund** [comment]
Note that this PR only sets position_ids and shifts the mask for **non seq2seq models**. I have only tried this on a gpt2 model, and is not sure whether this bug also applies to seq2seq models.

File: trlx/trainer/accelerate_ppo_trainer.py [hunk 1]
```diff
@@ -398,10 +408,21 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):  # noq
...
+                    1,
+                    keepdim=True,
+                )
+                position_ids = torch.max(
+                    torch.zeros_like(attention_mask),
+                    torch.arange(attention_mask.shape[1], device=attention_mask.device).repeat(
+                        (attention_mask.shape[0], 1)
+                    )
+                    - start_pos_idx,
+                )
```
Comment (maxreciprocate) [inline trlx/trainer/accelerate_ppo_trainer.py:422]: Do you think this computation can this be simplified down to:
```python
position_ids = attention_mask.long().cumsum(-1) - 1
position_ids.masked_fill_(attention_mask == 0, 1)
```
It seems to me that they are equivalent operationally (https://wandb.ai/sorry/trlx/reports/Different-position_ids-calculations--Vmlldzo0NzAzODg4)

**Review by maxreciprocate — CHANGES_REQUESTED** [review]
Thanks Tobias! This is an extremely valuable find

File: trlx/trainer/accelerate_ppo_trainer.py [hunk 2]
```diff
@@ -414,6 +435,7 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):  # noq
                         ref_logits = self.ref_model(
                             all_tokens,
                             attention_mask=attention_mask,
+                            position_ids=position_ids,
```
[the remark this answers is not in the document]
Comment (TobiasNorlund) [reply trlx/trainer/accelerate_ppo_trainer.py:438]: Should it also be duplicated to the `if self.config.model.model_arch_type == "seq2seq"` branches?

File: trlx/trainer/accelerate_ppo_trainer.py (hunk 1 above)
Comment (TobiasNorlund) [reply trlx/trainer/accelerate_ppo_trainer.py:422]: Yes that is a nicer way of doing it, thanks!

File: trlx/trainer/accelerate_ppo_trainer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxreciprocate) [reply trlx/trainer/accelerate_ppo_trainer.py:421]: Yes please!

[the remark this answers is not in the document]
Comment (TobiasNorlund) [reply trlx/trainer/accelerate_ppo_trainer.py:421]: Actually, I realize the `forward` methods for seq2seq models lack the `position_ids` argument, and at least T5 uses relative positional biases AFAIK, not absolute, in which case this should not be a problem for that model at least. I'm not sure whether there are other seq2seq models with absolute pos embeddings that TRLX support?

[the remark this answers is not in the document]
Comment (maxreciprocate) [reply trlx/trainer/accelerate_ppo_trainer.py:421]: You're correct, sorry for the confusion, there aren't any currently except T5

**Review by maxreciprocate — APPROVED** [review]
https://wandb.ai/sorry/trlx-references/reports/fix-logratio-bug-v-main--Vmlldzo0NzE2NzYw

Thanks again, Tobias
LGTM!
